src/surface/Surface.ts

```diff
@@ -0,0 +1,156 @@
+/**
+ * Represents the visible Surface of a Launchpad
+ *
+ * Allows read/write access to the currently displayed buttons
+ * on the Launchpad. At any point, the in-memory display buffer can
+ * be written to the Launchpad. Commands will only be sent to
+ * render the difference.
+ *
+ * Supports multiple layers, so that it is easy to show temporary colors
+ * (for example, while a button is being held down) and return to the
+ * original color unchanged afterwards. Setting colors on the Surface
+ * directly is the same as setting them on layer 0.
+ *
+ * Note that on higher layers, palette color 0 and 'off' are not the same: 'off'
+ * will make the underlying button show through, while palette color 0 will make
+ * the button actively turn off.
+ *
+ * There should only be one surface in use for any Launchpad
+ * at a given time.
+ */
+export class Surface implements ILayer {
+  public readonly width = 9;
+  public readonly height = 9;
+
+  private currentDisplay = new Layer(this.width, this.height);
+  private readonly layer0 = new Layer(this.width, this.height);
+  private readonly layers = new Map<number, Layer>();
+  private readonly coords: Array<[number, number]>;
+
+  constructor(private readonly lp: ILaunchpad) {
+    lp.allOff();
+    this.layers.set(0, this.layer0);
+
+    // Caching this array saves recalculating it every frame
+    this.coords = allXy(this.width, this.height);
+  }
+
+  /**
+   * Create or return the layer at the given index.
+   *
+   * Layer number must be a natural number.
+   */
+  public layer(i: number): Layer {
+    if (i < 0 || Math.floor(i) !== i) {
+      throw new Error(`Layer number must be a natural number, got: ${i}`);
+    }
+    let layer = this.layers.get(i);
+    if (!layer) {
+      layer = new Layer(this.width, this.height);
+      this.layers.set(i, layer);
+    }
+    return layer;
+  }
+
+  /**
+   * Remove a layer with the given number
+   *
+   * Layer 0 cannot be removed.
+   */
+  public removeLayer(i: number): void {
+    if (i < 0 || Math.floor(i) !== i) {
+      throw new Error(`Layer number must be a natural number, got: ${i}`);
+    }
+    if (i === 0) {
+      throw new Error('Cannot remove layer 0');
+    }
+    this.layers.delete(i);
+  }
+
+  /**
+   * Send the current display buffer to the Launchpad
+   */
+  public update(): void {
+    const newDisplay = this.flat();
+    const updates: ButtonStyle[] = [];
+
+    for (const xy of this.coords) {
+      const cur = this.currentDisplay.get(xy);
+      const updated = newDisplay.get(xy);
+      if (!sameStyle(cur, updated)) {
+        updates.push({
+          button: xy,
+          style: updated,
+        });
+      }
+    }
+
+    this.lp.setButtons(...updates);
+    this.currentDisplay = newDisplay;
+  }
+
+  set(x: number, y: number, style: Style): void;
+  set(xy: [number, number], style: Style): void;
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+  set(x: any, y: any, style?: any): void {
```

Comment:
why are these specified as "any" types?

Reply:
It's really the only way to write this in a way that TypeScript doesn't completely go bonkers over.

I *could* have written it like this:

```ts
set(x: number | [number, number], y: number | Style, style?: Style): void {
```

However, the implementation of this function still needs to type correctly. And when I then go to call `layer.set()`:

```ts
set(x: number | [number, number], y: number | Style, style?: Style): void {
  this.layer0.set(x, y, style);
//                ^^^^^^^^^^^ ERROR
//  Given the current types this could be this.layer0.set(number, Style, Style) for all the compiler knows
}
```

Which means I would need to do all the type checking in this function as well to rule out invalid combinations so that the compiler can type-check the implementation of this function.

Since I know the type checking is going to happen downstream, the easiest way to make the compiler shut up is to type everything as `any`.

Reply:
Wouldn't it just be better in that case to give `set(xy: [number, number], style: Style): void;` a body that calls the other function?

Reply:
That is annoying, guess it's being stuck with this